### PR TITLE
669 add matchlabels to template options selector

### LINKED ISSUE #669: Add matchLabels to template options selector
### Description of problem

The `matchFields` selector is not ideal for dealing with labels, so this adds a `matchLabels` selector.

### Proposed solution

Implement `matchLabels` and `matchExpressions` as per [this RFC](https://github.com/vmware-tanzu/cartographer/pull/591)

### Example

```yaml
  resources:
    - name: source-provider
      templateRef:
        kind: ClusterSourceTemplate
        options: 
        - name: imgpkg-bundle-template
          selector:
            matchLabels:                               # <========= add this
              app.tanzu.vmware.com/source-type: imgpkg-bundle
```

### PR DESCRIPTION
**Pull request by idoru**
## Changes proposed by this PR

This change introduces support for `matchLabels` and `matchExpressions` in a `TemplateOption`'s `selector`.

updates #669 

remaining to-do: more thorough validation testing in `cluster_supply_chain_validations_test` and `cluster_delivery_validations_test`.

Effort has been taken to ensure this behavior is consistent with the matching applied to owners (Workload/Deliverable) using their `selector`, `selectorMatchExpressions` and `selectorMatchFields` properties and to provide better clarity on those fields being a legacy design.

Also corrects behavior with top-level selector properties which improperly considered empty results from jsonpath evaluation of `matchFields` as an error.

## Release Note

`TemplateOption.Selector` can match labels of an owner using [`matchLabels` or `matchExpressions`](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/)

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [x] Removed non-atomic or `wip` commits
- [x] Filled in the [Release Note](#Release-Note) section above 
- [x] Modified the docs to match changes

### CODE REVIEW

File: pkg/selector/matcher.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+func (e selectorMatchError) SelectorIndex() int {
+	return e.SelectingObjectIndex
+}
+
+// BestSelectorMatchIndices returns a slice of indices into the passed in `selectors` which match the
+// `selectable` with the most specificity. Any error processing a selector includes the index of the
+// offending Selector.
+func BestSelectorMatchIndices(selectable Selectable, selectors []v1alpha1.Selector) ([]int, MatchError) {
+	var matchingSelectorIndicesByScore = map[int][]int{}
```
Comment (martyspiewak) [inline pkg/selector/matcher.go]: It seems like we're building out a whole map unnecessarily, no? Don't we really just want an array of the indices with the highest matching score? If we reach a new max score on line 88, we could trash the array and rebuild it.

I'm not too worried about the storage complexity, but I think it's a little less clear what's going on the way it's written right now.

File: pkg/repository/selector_matcher.go [hunk 2]
```diff
@@ -18,100 +18,60 @@ import (
...
 type SelectingObject interface {
-	GetSelectors() v1alpha1.Selectors
+	GetSelectors() v1alpha1.LegacySelector
 	GetObjectKind() schema.ObjectKind
 	GetName() string
 }
 
-type Selectable interface {
-	GetLabels() map[string]string
+func LegacySelectorToSelector(legacySelector v1alpha1.LegacySelector) v1alpha1.Selector {
```
Comment (martyspiewak) [inline pkg/repository/selector_matcher.go]: Can this method be private? Looks like it's only called on line 46

**Review by martyspiewak — APPROVED** [review]

**Review by martyspiewak — CHANGES_REQUESTED** [review]
Looks good to me! Left a couple of small comments but nothing big.

**Review by martyspiewak — APPROVED** [review]
Lgtm!

**Review by cirocosta — APPROVED** [review]

**Pull request merged**